main.cpp

```diff
@@ -75,23 +75,393 @@ write 3 UDTs below that EACH have:
+}
+
+int Dentist::costPerPatient(int num)
+{
+    for(int i = 1; i <= num; i++)
+    {
+        std::cout << "Patient " << i << "cost " << chargePerPatient - costDisposableEquipment << std::endl;
+    }
+    int profitPerWeek = costDisposableEquipment * num * 4;
+    return profitPerWeek;
+}
+
 /*
  copied UDT 3:
  */
 
+struct ATM
+{
+    ATM();
+    ~ATM()
+    {
+        std::cout << "ATM destructor called" << std::endl;
+    }
+    double currentAmountAvailable;
+    double maxAmountAvailable = 40000;
+    int numCustomers = 200;
+    int accountNumCurrentCustomer = 12345678;
+    int pinNumCurrentCustomer = 1234;
+
+    void dispenseCash();
+    int displayBalance(int accountNum); 
+    void dispenseReceipt(double amountWithdrawn);
+    double dispensingCash(double amountWithdrawn);
+};
+
+ATM::ATM()
+{
+    currentAmountAvailable = 20000;
+}
+
+void ATM::dispenseCash()
+{
+    std::cout << "Dispensing cash... " << std::endl;
+}
+
+int ATM::displayBalance(int accountNum)
+{
+    accountNumCurrentCustomer = accountNum;
+    std::cout << "Displaying balance for accout: " << accountNumCurrentCustomer << std::endl;
+    return accountNumCurrentCustomer;
+}
+
+void ATM::dispenseReceipt(double amountWithdrawn)
+{
+    currentAmountAvailable -= amountWithdrawn;
+    std::cout << "Dispensing receipt... " << std::endl;
+    std::cout << "Current amount available " << currentAmountAvailable << std::endl;
+}
+
+double ATM::dispensingCash(double amountWithdrawn)
+{
+    double amountDispensed = 0;
+    while (amountDispensed < amountWithdrawn)
+    {
+        amountDispensed += 10;
+    }
+    std::cout << "Dispensing requested amount" << std::endl;
+    maxAmountAvailable -= amountDispensed;
+    return maxAmountAvailable;
+}
+
 /*
  new UDT 4:
  with 2 member functions
  */
 
+struct Cafeteria
+{
+    Cafeteria()
+    {
+        coffeeMaker.switchedOn = true;
+        std::cout << "Cafeteria constructor called" << std::endl;
+    }
+    ~Cafeteria() 
+    { 
+        coffeeMaker.switchedOn = false;
+        std::cout << "Cafeteria destructor called" << std::endl;
+    } 
+
+    CoffeeMaker coffeeMaker;
+    CoffeeMaker::Cup cup1, cup2, cup3, cup4;
+
+    void serveCustomer(std::string type, std::string col);
+    float fillCoffeeMaker(); 
+};
+
+void Cafeteria::serveCustomer(std::string type, std::string col)
+{
+    cup1.colour = col;
+    coffeeMaker.typeOfCoffee = type;
+    std::cout << "Customer will receive " << coffeeMaker.typeOfCoffee << " in a " << cup1.colour  << " cup" << std::endl;
+}
+
+float Cafeteria::fillCoffeeMaker()
+{
+    if(coffeeMaker.waterRemaining < 250)
+    {
+        std::cout << "Filling reservoir " << std::endl;
+        coffeeMaker.waterRemaining = 1000;
+        return coffeeMaker.waterRemaining;
+    }
+    else
+    {
+        std::cout << "Reservoir does not need filling " << std::endl;
+        return coffeeMaker.waterRemaining;
+    }
```

Comment:
If you find yourself writing code like this:
```
int Type::someFunction()
{
    if( someExpression )
    {
        someStatement();
        return aValue;
    } 
    else 
    {
        return anotherValue;
    }
}
```
You do not need the `else` block.  Here's why:
`return` statements are the last thing that **ever happens** in a function.
When the `if()` expression is evaluated as `true` and the block is entered, the `return aValue;` is the last line of code to be executed before the function exits.
but when the `if()` expression evaluates as `false` and the block is **NOT** entered, the next thing that happens is `return anotherValue;`
you have no code after the `else` block that needs executing, since your `else` block contains a `return` statement.
that means you do **NOT** need the `else` block, and it is just causing visual clutter.
```
int Type::someFunction()
{
    if( someExpression )
    {
        someStatement();
        return aValue;
    } 
    
    return anotherValue;
}
```
- [ ] revise

Reply:
Thanks -- I keep making this mistake.